com.unity.render-pipelines.universal/Runtime/LightCookieManager.cs

```diff
@@ -534,119 +665,228 @@ int FilterAndValidateAdditionalLights(ref LightData lightData, LightCookieMappin
+                        m_CookieSizeDivisor++;
+                        m_PrevCookieRequestPixelCount = cookieRequestPixelCount;
+#if VERBOSE_DEBUG
+                        Debug.Log($"Increase cookie divisor to: {m_CookieSizeDivisor}");
+#endif
+                    }
+                    else
+                    {
+                        // Reduce cookie size to approximate value try to rebuild.
+                        m_CookieSizeDivisor = Mathf.Max(m_CookieSizeDivisor + 1, cookieSizeDivisorApprox);
+                        m_PrevCookieRequestPixelCount = cookieRequestPixelCount;
+#if VERBOSE_DEBUG
+                        Debug.Log($"Atlas full, increase divisor and try again: {m_CookieSizeDivisor}");
+#endif
                     }
 
+                    // Clear atlas and try to rebuild it.
+                    // NOTE:
+
                     // Clear atlas allocs
                     m_AdditionalLightsCookieAtlas.ResetAllocator();
+                    atlasReset = true;
+#if VERBOSE_DEBUG
+                    Debug.Log($"Atlas reset, try again divisor: {m_CookieSizeDivisor}");
+#endif
 
                     // Try to reinsert in priority order
-                    i = 0;
                     uvRectCount = 0;
-                    atlasResetBefore = true;
+                    i = -1; // Incremented right after continue
                     continue;
                 }
 
                 // Adjust atlas UVs for OpenGL
                 if (!SystemInfo.graphicsUVStartsAtTop)
                     uvScaleOffset.w = 1.0f - uvScaleOffset.w - uvScaleOffset.y;
+#if VERBOSE_DEBUG
+                Debug.Log($"i: {i}, '{cookie.name}' {cookie.width}x{cookie.height} Rect: {uvScaleOffset}");
+#endif
 
-                textureAtlasUVRects[uvRectCount++] = uvScaleOffset;
+                uvRectCount++;
+                textureAtlasUVRects[lcm.lightBufferIndex] = uvScaleOffset;
             }
 
+            // Restore linear buffer order for shader data setup and upload.
+            // (Alternatively, we could use the mappings to do scattered read/writes).
+            validLightMappings.Sort(LightCookieMapping.s_CompareByBufferIndex);
+
             return uvRectCount;
         }
 
-        Vector4 Fetch2D(CommandBuffer cmd, Texture cookie)
+        uint ComputeCookieRequestPixelCount(ref WorkSlice<LightCookieMapping> validLightMappings)
+        {
+            uint requestPixelCount = 0;
+            m_UniqueCookieTextureIDs.Clear();
+            for (int i = 0; i < validLightMappings.length; i++)
+            {
+                var lcm = validLightMappings[i];
+                //Light light = lightData.visibleLights[lcm.visibleLightIndex].light;
+                Texture cookie = lcm.legacyLight.cookie;
+                int cookieID = cookie.GetInstanceID();
+
+                // Consider only unique textures as atlas request pixels
+                if (m_UniqueCookieTextureIDs.Contains(cookieID))
```

Comment:
I realized afterwards that, now that we sort the cookies, we could easily check duplicates without the map if we'd sort by size and textureID but perhaps that's a fix for another day.